# **ML APPROACH**

- feature engineering for pooled gradient
boosting model

**LOADING AND CLEANING COMMODITY DATA**

In [1]:
# Libraries

from pathlib import Path
import pandas as pd
import numpy as np
from sklearn.model_selection import LeaveOneGroupOut

In [2]:
print(Path.cwd())

C:\Users\mason\OneDrive\Desktop\AgGenius\wasde_analysis\notebooks\models


In [3]:
PROJECT_ROOT = Path.cwd().parents[1]
DATA_DIR = PROJECT_ROOT / 'data' / 'processed' / 'commodity_prod'

print(Path.cwd())
print(DATA_DIR.exists())

C:\Users\mason\OneDrive\Desktop\AgGenius\wasde_analysis\notebooks\models
True


In [4]:
commodity_files = {
    'Corn': 'corn_prod.csv',
    'Cotton': 'cott_prod.csv',
    'Hard Red Spring Wheat': 'hrs_prod.csv',
    'Hard Red Winter Wheat': 'hrw_prod.csv',
    'Oats': 'oats_prod.csv',
    'Rice (Rough)': 'rr_prod.csv',
    'Soybean Meal': 'soy_meal_prod.csv',
    'Soybean Oil': 'soy_oil_prod.csv',
    'Soybeans': 'soybean_prod (2).csv',
    'Sugar': 'sugar_prod.csv',
    'Wheat': 'wht_prod.csv',
}

raw_data = {}
for commodity, filename in commodity_files.items():
    path = DATA_DIR / filename
    raw_data[commodity] = pd.read_csv(path, index_col=0, parse_dates=False)
    print(commodity, '->', raw_data[commodity].shape)

Corn -> (567, 15)
Cotton -> (567, 15)
Hard Red Spring Wheat -> (346, 15)
Hard Red Winter Wheat -> (346, 15)
Oats -> (567, 15)
Rice (Rough) -> (567, 15)
Soybean Meal -> (567, 15)
Soybean Oil -> (567, 15)
Soybeans -> (567, 15)
Sugar -> (567, 15)
Wheat -> (567, 15)


In [5]:
for commodity, d in raw_data.items():
    if 'ReliabilityProjection' in d.columns:
        d.drop(columns='ReliabilityProjection', inplace=True)
    if d['ReportDate'].dtype == str:
        d['ReportDate'] = pd.to_datetime(d['ReportDate'])

# Confirm the cleanup landed everywhere
for commodity, d in raw_data.items():
    print(commodity, d.shape, d['ReportDate'].dtype)

Corn (567, 14) str
Cotton (567, 14) str
Hard Red Spring Wheat (346, 14) str
Hard Red Winter Wheat (346, 14) str
Oats (567, 14) str
Rice (Rough) (567, 14) str
Soybean Meal (567, 14) str
Soybean Oil (567, 14) str
Soybeans (567, 14) str
Sugar (567, 14) str
Wheat (567, 14) str


In [6]:
print(f"Loaded {len(raw_data)} of {len(commodity_files)} expected commodities")
missing = set(commodity_files.keys()) - set(raw_data.keys())
print('Missing:', missing if missing else 'none')

Loaded 11 of 11 expected commodities
Missing: none


In [7]:
excluded_commodities = ['Hard Red Spring Wheat', 'Hard Red Winter Wheat']

raw_data_modeling = {k: v for k, v in raw_data.items() if k not in excluded_commodities}

print(f"Modeling on {len(raw_data_modeling)} commodities: {list(raw_data_modeling.keys())}")

Modeling on 9 commodities: ['Corn', 'Cotton', 'Oats', 'Rice (Rough)', 'Soybean Meal', 'Soybean Oil', 'Soybeans', 'Sugar', 'Wheat']


DateTime conversion

In [8]:
for commodity, d in raw_data_modeling.items():
    if not pd.api.types.is_datetime64_any_dtype(d['ReportDate']):
        d['ReportDate'] = pd.to_datetime(d['ReportDate'])
    print(commodity, d['ReportDate'].dtype)

Corn datetime64[us]
Cotton datetime64[us]
Oats datetime64[us]
Rice (Rough) datetime64[us]
Soybean Meal datetime64[us]
Soybean Oil datetime64[us]
Soybeans datetime64[us]
Sugar datetime64[us]
Wheat datetime64[us]


C:\Users\mason\AppData\Local\Temp\ipykernel_8460\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_8460\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_8460\2340286167.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  d['ReportDate'] = pd.to_datetime(d['ReportDate'])
C:\Users\mason\AppData\Local\Temp\ipykernel_8460\2340286167.py:3: UserWarning: Could not infer

**Excluded:** Hard Red Spring Wheat, Hard Red Winter Wheat

WASDE never publishes a finalized (ProjEstFlag = NaN) production figure at the wheat-subclass level — confirmed across both raw commodity files and the master concatenated source, under every label variant. 

Only aggregate 'Wheat' production is ever finalized. Without ground truth, these two can't be validated as standalone nowcasting targets. 

**Possible future extension:** use HRS/HRW live estimates as input features for predicting total Wheat's finalized value, rather than as targets themselves

# **Run 'build_vintages' function across all commodities**

In [9]:
def build_vintages(df, commodity_name=None):
    """
    Given a cleaned WASDE dataframe for one commodity, build the vintage
    table: every live (Proj./Est.) estimate tagged with its horizon to
    finalization and the eventual final value.
    """
    df = df.copy()

    # First-finalized appearance per market year
    finals_info = (
        df[df['ProjEstFlag'].isna()]
          .sort_values('ReportDate')
          .groupby('MarketYear')
          .head(1)
          .set_index('MarketYear')[['ReportDate', 'Value']]
          .rename(columns={'ReportDate': 'final_date', 'Value': 'final_value'})
    )

    vintages = df[df['ProjEstFlag'].notna()].copy()
    vintages['final_value'] = vintages['MarketYear'].map(finals_info['final_value'])
    vintages['final_date'] = vintages['MarketYear'].map(finals_info['final_date'])

    vintages['horizon_months'] = (
        (vintages['final_date'].dt.year - vintages['ReportDate'].dt.year) * 12
        + (vintages['final_date'].dt.month - vintages['ReportDate'].dt.month)
    )

    # Drop any market year without a confirmed final value (still-live years)
    vintages = vintages.dropna(subset=['final_value', 'final_date']).copy()

    if commodity_name:
        vintages['Commodity'] = commodity_name

    return vintages, finals_info

In [10]:
all_vintages = []
finals_by_commodity = {}

for commodity, d in raw_data_modeling.items():
    v, finals_info = build_vintages(d, commodity_name=commodity)
    all_vintages.append(v)
    finals_by_commodity[commodity] = finals_info
    print(f"{commodity}: {len(v)} vintage rows, {v['MarketYear'].nunique()} finalized market years, max horizon {v['horizon_months'].max()}")

master_vintages = pd.concat(all_vintages, ignore_index=True)
print('\nTotal pooled rows:', master_vintages.shape)

Corn: 344 vintage rows, 16 finalized market years, max horizon 24.0


Cotton: 322 vintage rows, 16 finalized market years, max horizon 24.0
Oats: 344 vintage rows, 16 finalized market years, max horizon 24.0
Rice (Rough): 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Meal: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Oil: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybeans: 344 vintage rows, 16 finalized market years, max horizon 24.0
Sugar: 344 vintage rows, 16 finalized market years, max horizon 24.0
Wheat: 344 vintage rows, 16 finalized market years, max horizon 24.0

Total pooled rows: (3074, 17)


In [11]:
for commodity, d in raw_data_modeling.items():
    n_null_flag = d['ProjEstFlag'].isna().sum()
    unique_flags = d['ProjEstFlag'].unique()
    print(f"{commodity}: {n_null_flag} finalized rows | flags: {list(unique_flags)}")

Corn: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Cotton: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Oats: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Rice (Rough): 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybean Meal: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybean Oil: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Soybeans: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Sugar: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']
Wheat: 189 finalized rows | flags: [nan, 'Est.', 'Proj.']


# **ESTABLISHING A BASELINE**

In [12]:
master_vintages['raw_error'] = master_vintages['final_value'] - master_vintages['Value']
master_vintages['raw_abs_error'] = master_vintages['raw_error'].abs()

master_vintages[['Commodity', 'MarketYear', 'horizon_months', 'Value', 'final_value', 'raw_error']].head(10)

,Commodity,MarketYear,horizon_months,Value,final_value,raw_error
0,Corn,2008/09,1.0,12092.0,12092.0,0.0
1,Corn,2009/10,13.0,13131.0,13092.0,-39.0
2,Corn,2009/10,12.0,13110.0,13092.0,-18.0
3,Corn,2010/11,24.0,13370.0,12447.0,-923.0
4,Corn,2009/10,11.0,13110.0,13092.0,-18.0
5,Corn,2010/11,23.0,13370.0,12447.0,-923.0
6,Corn,2009/10,10.0,13110.0,13092.0,-18.0
7,Corn,2010/11,22.0,13245.0,12447.0,-798.0
8,Corn,2009/10,9.0,13110.0,13092.0,-18.0
9,Corn,2010/11,21.0,13365.0,12447.0,-918.0


In [13]:
baseline_by_commodity_horizon = (
    master_vintages.groupby(['Commodity', 'horizon_months'])['raw_abs_error']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'mae_raw', 'count': 'n'})
    .reset_index()
)

print(baseline_by_commodity_horizon.head(20))

   Commodity  horizon_months     mae_raw   n
0       Corn             1.0    0.000000  16
1       Corn             2.0    0.000000  15
2       Corn             3.0    0.000000  15
3       Corn             4.0    0.000000  14
4       Corn             5.0   17.733333  15
5       Corn             6.0   17.733333  15
6       Corn             7.0   20.461538  13
7       Corn             8.0   26.466667  15
8       Corn             9.0   26.466667  15
9       Corn            10.0   26.466667  15
10      Corn            11.0   26.466667  15
11      Corn            12.0   29.133333  15
12      Corn            13.0   32.466667  15
13      Corn            14.0   32.000000  14
14      Corn            15.0   32.000000  14
15      Corn            16.0   28.307692  13
16      Corn            17.0  130.785714  14
17      Corn            18.0  130.785714  14
18      Corn            19.0  219.166667  12
19      Corn            20.0  259.000000  14


In [14]:
baseline_by_horizon_pooled = (
    master_vintages.groupby('horizon_months')['raw_abs_error']
    .agg(['mean', 'count'])
    .rename(columns={'mean': 'mae_raw', 'count': 'n'})
)

print(baseline_by_horizon_pooled)

                   mae_raw    n
horizon_months                 
1.0               0.152778  144
2.0               0.222222  135
3.0               1.600000  135
4.0               1.634921  126
5.0               3.911111  135
6.0               4.392593  135
7.0              28.237607  117
8.0              41.477612  134
9.0              66.149254  134
10.0             82.350746  134
11.0            105.485075  134
12.0            140.037313  134
13.0            168.506418  134
14.0            193.076960  125
15.0            187.557280  125
16.0            204.803793  116
17.0            226.640480  125
18.0            242.295680  125
19.0            294.460935  107
20.0            316.090000  124
21.0            317.107823  124
22.0            304.650484  124
23.0            363.614032  124
24.0            379.542258  124


In [15]:
master_vintages['raw_pct_error'] = (master_vintages['raw_error'] / master_vintages['final_value']).abs() * 100

baseline_by_commodity_horizon_full = (
    master_vintages.groupby(['Commodity', 'horizon_months'])
    .agg(
        mae_raw=('raw_abs_error', 'mean'),
        mape_raw=('raw_pct_error', 'mean'),
        n=('raw_abs_error', 'count')
    )
    .reset_index()
)

print(baseline_by_commodity_horizon_full.head(20))

   Commodity  horizon_months     mae_raw  mape_raw   n
0       Corn             1.0    0.000000  0.000000  16
1       Corn             2.0    0.000000  0.000000  15
2       Corn             3.0    0.000000  0.000000  15
3       Corn             4.0    0.000000  0.000000  14
4       Corn             5.0   17.733333  0.127743  15
5       Corn             6.0   17.733333  0.127743  15
6       Corn             7.0   20.461538  0.147396  13
7       Corn             8.0   26.466667  0.188648  15
8       Corn             9.0   26.466667  0.188648  15
9       Corn            10.0   26.466667  0.188648  15
10      Corn            11.0   26.466667  0.188648  15
11      Corn            12.0   29.133333  0.208227  15
12      Corn            13.0   32.466667  0.233116  15
13      Corn            14.0   32.000000  0.228489  14
14      Corn            15.0   32.000000  0.228489  14
15      Corn            16.0   28.307692  0.203151  13
16      Corn            17.0  130.785714  0.942466  14
17      Co

In [16]:
mape_by_commodity = (
    master_vintages.groupby('Commodity')['raw_pct_error']
    .mean()
    .sort_values()
)
print(mape_by_commodity)

Commodity
Wheat           0.777000
Corn            1.134595
Soybeans        1.331023
Rice (Rough)    1.429008
Soybean Meal    1.435916
Soybean Oil     1.447467
Sugar           1.618955
Cotton          2.561545
Oats            2.602168
Name: raw_pct_error, dtype: float64


- Wheat is the most predictable (0.78% average error) — USDA's live estimates for aggregate wheat production are already very close to final, across all horizons combined.
- Corn, Soybeans, Rice, Soybean Meal/Oil cluster in a middle band (1.1-1.4%) — decent, but with more room for a model to find real signal.
- Sugar (1.6%), Cotton (2.6%), Oats (2.6%) are the least predictable — WASDE's own early-season numbers for these are proportionally the roughest.

**LOCKED IN BASELINE**

In [17]:
baseline_by_commodity_horizon_full.to_csv(DATA_DIR / 'raw_wasde_baseline.csv', index=False)
print("Baseline saved.")

Baseline saved.


# **FEATURE ENGINEERING**

**Commodity**

In [18]:
master_vintages['Commodity'] = master_vintages['Commodity'].astype('category')
print(master_vintages['Commodity'].dtype)
print(master_vintages['Commodity'].cat.categories.tolist())

category
['Corn', 'Cotton', 'Oats', 'Rice (Rough)', 'Soybean Meal', 'Soybean Oil', 'Soybeans', 'Sugar', 'Wheat']


**Revision trajectory (how much the estimates moved from previous report)**

In [19]:
master_vintages = master_vintages.sort_values(['Commodity', 'MarketYear', 'ReportDate']).copy()

master_vintages['prior_value'] = (
    master_vintages.groupby(['Commodity', 'MarketYear'])['Value'].shift(1)
)

master_vintages['revision'] = master_vintages['Value'] - master_vintages['prior_value']
master_vintages['revision_pct'] = master_vintages['revision'] / master_vintages['prior_value'] * 100

master_vintages[['Commodity', 'MarketYear', 'ReportDate', 'Value', 'prior_value', 'revision', 'revision_pct']].head(15)

,Commodity,MarketYear,ReportDate,Value,prior_value,revision,revision_pct
0,Corn,2008/09,2010-04-01,12092.0,NaN,NaN,NaN
1,Corn,2009/10,2010-04-01,13131.0,NaN,NaN,NaN
2,Corn,2009/10,2010-05-01,13110.0,13131.0,-21.0,-0.159927
4,Corn,2009/10,2010-06-01,13110.0,13110.0,0.0,0.000000
6,Corn,2009/10,2010-07-01,13110.0,13110.0,0.0,0.000000
8,Corn,2009/10,2010-08-01,13110.0,13110.0,0.0,0.000000
11,Corn,2009/10,2010-09-01,13110.0,13110.0,0.0,0.000000
12,Corn,2009/10,2010-10-01,13110.0,13110.0,0.0,0.000000
14,Corn,2009/10,2010-11-01,13110.0,13110.0,0.0,0.000000
17,Corn,2009/10,2010-12-01,13110.0,13110.0,0.0,0.000000


In [20]:
print(master_vintages['revision'].isna().sum())
print(master_vintages['revision'].isna().sum() / len(master_vintages) * 100, '%')

144
4.68445022771633 %


In [21]:
master_vintages['is_first_estimate'] = master_vintages['prior_value'].isna().astype(int)
master_vintages['revision'] = master_vintages['revision'].fillna(0)
master_vintages['revision_pct'] = master_vintages['revision_pct'].fillna(0)

master_vintages[['Commodity', 'MarketYear', 'ReportDate', 'revision', 'revision_pct', 'is_first_estimate']].head(5)

,Commodity,MarketYear,ReportDate,revision,revision_pct,is_first_estimate
0,Corn,2008/09,2010-04-01,0.0,0.000000,1
1,Corn,2009/10,2010-04-01,0.0,0.000000,1
2,Corn,2009/10,2010-05-01,-21.0,-0.159927,0
4,Corn,2009/10,2010-06-01,0.0,0.000000,0
6,Corn,2009/10,2010-07-01,0.0,0.000000,0


**Report Month - Seasonality**

In [22]:
master_vintages['report_month'] = master_vintages['ReportDate'].dt.month

master_vintages.groupby('report_month').size().sort_index()

report_month
1     242
2     260
3     260
4     278
5     258
6     258
7     258
8     258
9     258
10    224
11    260
12    260
dtype: int64

In [23]:
master_vintages['report_month'] = master_vintages['report_month'].astype('category')

**Normalized version**

- Expresses each estimate relative to that commoditys own historical typical value

In [24]:
commodity_stats = (
    master_vintages.groupby('Commodity')['final_value']
    .agg(['mean', 'std'])
    .rename(columns={'mean': 'commodity_mean', 'std': 'commodity_std'})
)

master_vintages = master_vintages.merge(commodity_stats, on='Commodity', how='left')

master_vintages['value_zscore'] = (
    (master_vintages['Value'] - master_vintages['commodity_mean']) / master_vintages['commodity_std']
)

master_vintages[['Commodity', 'Value', 'commodity_mean', 'commodity_std', 'value_zscore']].groupby('Commodity').first()

,Value,commodity_mean,commodity_std,value_zscore
Commodity,,,,
Corn,12092.00,13764.758721,1184.069726,-1.412720
Cotton,12.82,16.106801,2.695367,-1.219426
Oats,89.00,63.433140,13.460932,1.899338
Rice (Rough),203.70,203.795640,22.465554,-0.004257
Soybean Meal,39104.00,46458.421512,4944.040844,-1.487533
Soybean Oil,18746.00,22813.078488,2671.251689,-1.522537
Soybeans,2967.00,3866.055233,498.922442,-1.801994
Sugar,7532.00,8801.069767,470.201137,-2.698993
Wheat,2499.00,1976.075581,212.129863,2.465115


What this does: 

for each commodity, compute the mean and standard deviation of its final (true) production values across all market years, 

then express every live estimate as a z-score relative to that — "how many standard deviations above/below this commodity's typical production level is this estimate." 

Now a value of 2.0 means the same thing (unusually high) whether it's Corn or Cotton, which the raw Value alone can't tell you.

**CONSOLIDATING THE FEATURE SET**

In [25]:
feature_cols = [
    'Commodity', 'horizon_months', 'revision', 'revision_pct',
    'is_first_estimate', 'report_month', 'value_zscore'
]
target_col = 'final_value'

print(master_vintages[feature_cols + [target_col]].isna().sum())
print()
print(master_vintages[feature_cols].dtypes)

Commodity            0
horizon_months       0
revision             0
revision_pct         0
is_first_estimate    0
report_month         0
value_zscore         0
final_value          0
dtype: int64

Commodity            category
horizon_months        float64
revision              float64
revision_pct          float64
is_first_estimate       int64
report_month         category
value_zscore          float64
dtype: object


In [26]:
feature_cols = [
    'Commodity', 'horizon_months', 'Value', 'value_zscore',
    'revision', 'revision_pct', 'is_first_estimate', 'report_month'
]

print(master_vintages[feature_cols].dtypes)

Commodity            category
horizon_months        float64
Value                 float64
value_zscore          float64
revision              float64
revision_pct          float64
is_first_estimate       int64
report_month         category
dtype: object


# **LEAVE-ONE-MARKET-YEAR-OUT SPLIT STRATEGY**

Since value_zscore depends on commodity_mean/commodity_std, and those were computed from final_value across all years, 

we need a version that recomputes them fold-by-fold using only the training years:

In [27]:
def add_zscore_feature(train_df, test_df):
    """
    Compute commodity mean/std from train_df only, then apply that
    same normalization to both train_df and test_df. Prevents the
    held-out year's own final_value from leaking into its own
    normalization.
    """
    stats = (
        train_df.groupby('Commodity')['final_value']
        .agg(['mean', 'std'])
        .rename(columns={'mean': 'commodity_mean', 'std': 'commodity_std'})
    )

    train_df = train_df.merge(stats, on='Commodity', how='left')
    test_df = test_df.merge(stats, on='Commodity', how='left')

    train_df['value_zscore'] = (train_df['Value'] - train_df['commodity_mean']) / train_df['commodity_std']
    test_df['value_zscore'] = (test_df['Value'] - test_df['commodity_mean']) / test_df['commodity_std']

    return train_df, test_df

In [28]:
cols_to_drop = ['commodity_mean', 'commodity_std', 'value_zscore']
master_vintages = master_vintages.drop(columns=[c for c in cols_to_drop if c in master_vintages.columns])

print(master_vintages.columns.tolist())

['ReportDate', 'ReportTitle', 'Attribute', 'Commodity', 'Region', 'MarketYear', 'ProjEstFlag', 'AnnualQuarterFlag', 'Value', 'Unit', 'ReleaseDate', 'ReleaseTime', 'ForecastYear', 'ForecastMonth', 'final_value', 'final_date', 'horizon_months', 'raw_error', 'raw_abs_error', 'raw_pct_error', 'prior_value', 'revision', 'revision_pct', 'is_first_estimate', 'report_month']


In [29]:
feature_cols_base = [
    'Commodity', 'horizon_months', 'value_zscore',
    'revision', 'revision_pct', 'is_first_estimate', 'report_month'
]
target_col = 'final_value'

# 'Value' dropped: it's collinear with value_zscore by construction
# (zscore is a per-commodity linear transform of Value), so keeping both
# adds no signal for a tree model and just muddies feature importances.

In [30]:
cotton = master_vintages[master_vintages['Commodity'] == 'Cotton']
print(sorted(cotton['MarketYear'].unique()))

overlap_check = cotton[cotton['MarketYear'].isin(['2010/11', '2010/011'])]
print(overlap_check[['MarketYear', 'ReportDate', 'Value']].sort_values('ReportDate'))

others = master_vintages[(master_vintages['MarketYear'].str.contains('/0')) & (master_vintages['Commodity'] != 'Cotton')]
print(others[['Commodity', 'MarketYear', 'ReportDate', 'Value']])

['2008/09', '2009/10', '2010/011', '2011/012', '2012/13', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24']
    MarketYear ReportDate  Value
358   2010/011 2010-05-01  16.70
359   2010/011 2010-06-01  16.70
360   2010/011 2010-07-01  18.30
361   2010/011 2010-08-01  18.53
362   2010/011 2010-09-01  18.84
363   2010/011 2010-10-01  18.87
364   2010/011 2010-11-01  18.42
365   2010/011 2010-12-01  18.27
366   2010/011 2011-01-01  18.32
367   2010/011 2011-02-01  18.32
368   2010/011 2011-03-01  18.32
369   2010/011 2011-04-01  18.10
370   2010/011 2011-05-01  18.10
371   2010/011 2011-06-01  18.10
372   2010/011 2011-07-01  18.10
373   2010/011 2011-08-01  18.10
374   2010/011 2011-09-01  18.10
375   2010/011 2011-10-01  18.10
376   2010/011 2011-11-01  18.10
377   2010/011 2011-12-01  18.10
378   2010/011 2012-01-01  18.10
379   2010/011 2012-02-01  18.10
380   2010/011 2012-03-01  18.10
381   2010/011 2012-04-01  18

In [31]:
# Fix Cotton's mislabeled market years at the source, before rebuilding vintages
raw_data_modeling['Cotton']['MarketYear'] = raw_data_modeling['Cotton']['MarketYear'].replace({
    '2010/011': '2010/11',
    '2011/012': '2011/12'
})

# Confirm the fix landed
print(sorted(raw_data_modeling['Cotton']['MarketYear'].unique()))

['2007/08', '2008/09', '2009/10', '2010/11', '2011/12', '2012/013', '2012/13', '2013/014', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24', '2024/25', '2025/26']


In [32]:
import re

cotton_years = sorted(raw_data_modeling['Cotton']['MarketYear'].unique())
malformed = [y for y in cotton_years if re.match(r'^\d{4}/0\d{2}$', y)]
print("Malformed years found:", malformed)

Malformed years found: ['2012/013', '2013/014']


In [33]:
def fix_market_year(y):
    m = re.match(r'^(\d{4})/0(\d{2})$', y)
    if m:
        return f"{m.group(1)}/{m.group(2)}"
    return y

raw_data_modeling['Cotton']['MarketYear'] = raw_data_modeling['Cotton']['MarketYear'].apply(fix_market_year)

# Confirm no malformed years remain
remaining = [y for y in raw_data_modeling['Cotton']['MarketYear'].unique() if re.match(r'^\d{4}/0\d{2}$', y)]
print("Remaining malformed:", remaining)
print(sorted(raw_data_modeling['Cotton']['MarketYear'].unique()))

Remaining malformed: []
['2007/08', '2008/09', '2009/10', '2010/11', '2011/12', '2012/13', '2013/14', '2014/15', '2015/16', '2016/17', '2017/18', '2018/19', '2019/20', '2020/21', '2021/22', '2022/23', '2023/24', '2024/25', '2025/26']


In [34]:
all_vintages = []
finals_by_commodity = {}

for commodity, d in raw_data_modeling.items():
    v, finals_info = build_vintages(d, commodity_name=commodity)
    all_vintages.append(v)
    finals_by_commodity[commodity] = finals_info
    print(f"{commodity}: {len(v)} vintage rows, {v['MarketYear'].nunique()} finalized market years, max horizon {v['horizon_months'].max()}")

master_vintages = pd.concat(all_vintages, ignore_index=True)
print('\nTotal pooled rows:', master_vintages.shape)

Corn: 344 vintage rows, 16 finalized market years, max horizon 24.0
Cotton: 344 vintage rows, 16 finalized market years, max horizon 24.0
Oats: 344 vintage rows, 16 finalized market years, max horizon 24.0
Rice (Rough): 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Meal: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybean Oil: 344 vintage rows, 16 finalized market years, max horizon 24.0
Soybeans: 344 vintage rows, 16 finalized market years, max horizon 24.0


Sugar: 344 vintage rows, 16 finalized market years, max horizon 24.0
Wheat: 344 vintage rows, 16 finalized market years, max horizon 24.0

Total pooled rows: (3096, 17)


In [35]:
# Report month
master_vintages['report_month'] = master_vintages['ReportDate'].dt.month
master_vintages['report_month'] = master_vintages['report_month'].astype('category')

# Revision trajectory
master_vintages = master_vintages.sort_values(['Commodity', 'MarketYear', 'ReportDate']).copy()
master_vintages['prior_value'] = master_vintages.groupby(['Commodity', 'MarketYear'])['Value'].shift(1)
master_vintages['revision'] = master_vintages['Value'] - master_vintages['prior_value']
master_vintages['revision_pct'] = master_vintages['revision'] / master_vintages['prior_value'] * 100
master_vintages['is_first_estimate'] = master_vintages['prior_value'].isna().astype(int)
master_vintages['revision'] = master_vintages['revision'].fillna(0)
master_vintages['revision_pct'] = master_vintages['revision_pct'].fillna(0)

# Commodity as category
master_vintages['Commodity'] = master_vintages['Commodity'].astype('category')

# Raw + pct error vs baseline (for later comparison)
master_vintages['raw_error'] = master_vintages['final_value'] - master_vintages['Value']
master_vintages['raw_abs_error'] = master_vintages['raw_error'].abs()
master_vintages['raw_pct_error'] = (master_vintages['raw_error'] / master_vintages['final_value']).abs() * 100

print(master_vintages.shape)
print(master_vintages.isna().sum().sum(), 'total NaNs')

(3096, 25)
144 total NaNs


In [36]:
print(master_vintages[['revision', 'revision_pct']].isna().sum())
print(master_vintages['prior_value'].isna().sum())

revision        0
revision_pct    0
dtype: int64
144


In [37]:
cols_to_drop = ['commodity_mean', 'commodity_std', 'value_zscore']
master_vintages = master_vintages.drop(columns=[c for c in cols_to_drop if c in master_vintages.columns])

logo = LeaveOneGroupOut()
groups = master_vintages['MarketYear']

fold_results = []

for train_idx, test_idx in logo.split(master_vintages, groups=groups):
    train_raw = master_vintages.iloc[train_idx].copy()
    test_raw = master_vintages.iloc[test_idx].copy()

    train_fold, test_fold = add_zscore_feature(train_raw, test_raw)

    held_out_year = test_fold['MarketYear'].iloc[0]

    fold_results.append({
        'held_out_year': held_out_year,
        'n_train': len(train_fold),
        'n_test': len(test_fold)
    })

fold_summary = pd.DataFrame(fold_results)
print(fold_summary)
print('\nTotal folds:', len(fold_summary))

   held_out_year  n_train  n_test
0        2008/09     3087       9
1        2009/10     2979     117
2        2010/11     2880     216
3        2011/12     2880     216
4        2012/13     2889     207
5        2013/14     2889     207
6        2014/15     2880     216
7        2015/16     2880     216
8        2016/17     2880     216
9        2017/18     2889     207
10       2018/19     2889     207
11       2019/20     2880     216
12       2020/21     2880     216
13       2021/22     2889     207
14       2022/23     2889     207
15       2023/24     2880     216



Total folds: 16


# **FIT POOLED GRADIENT BOOSTING MODEL**

LightGBM or XGBoost on the full 3,074-row pooled set. Start simple (few features, shallow trees) before adding complexity.

In [38]:
import lightgbm as lgb

oof_rows = []

for train_idx, test_idx in logo.split(master_vintages, groups=groups):
    train_raw = master_vintages.iloc[train_idx].copy()
    test_raw = master_vintages.iloc[test_idx].copy()

    train_fold, test_fold = add_zscore_feature(train_raw, test_raw)

    # Target the residual (final_value - Value), not the raw level.
    # final_value is an unbounded, trending quantity that ranges from ~12
    # (Cotton) to ~46,000 (Soybean Meal) -- a tree ensemble can't extrapolate
    # past the target range it saw in training, so on held-out years with
    # record highs/lows the leaf-average predictions clamp badly. The
    # residual is small and roughly stationary across commodities/years,
    # which trees handle far better. Value itself stays out of the feature
    # set (collinear with value_zscore) -- it's only used here to
    # reconstruct the final prediction.
    train_target = train_fold[target_col] - train_fold['Value']

    model = lgb.LGBMRegressor(
        n_estimators=200,
        max_depth=4,
        learning_rate=0.05,
        random_state=42,
        verbosity=-1
    )
    model.fit(train_fold[feature_cols_base], train_target)

    test_fold = test_fold.copy()
    test_fold['ml_pred'] = test_fold['Value'] + model.predict(test_fold[feature_cols_base])

    oof_rows.append(
        test_fold[['Commodity', 'MarketYear', 'horizon_months', 'final_value', 'ml_pred']]
    )

oof = pd.concat(oof_rows, ignore_index=True)
print(oof.shape)
oof.head(10)

(3096, 5)


,Commodity,MarketYear,horizon_months,final_value,ml_pred
0,Corn,2008/09,1.0,12092.00,11954.614026
1,Cotton,2008/09,1.0,12.82,-25.212953
2,Oats,2008/09,1.0,89.00,75.349917
3,Rice (Rough),2008/09,1.0,203.70,228.608195
4,Soybean Meal,2008/09,1.0,39104.00,39087.927848
5,Soybean Oil,2008/09,1.0,18746.00,18739.412897
6,Soybeans,2008/09,1.0,2967.00,2963.359369
7,Sugar,2008/09,1.0,7532.00,7594.588510
8,Wheat,2008/09,1.0,2499.00,2485.349917
9,Corn,2009/10,13.0,13092.00,13065.689333


In [39]:
oof['ml_error'] = oof['final_value'] - oof['ml_pred']
oof['ml_abs_error'] = oof['ml_error'].abs()
oof['ml_pct_error'] = (oof['ml_error'] / oof['final_value']).abs() * 100

print(oof[['ml_abs_error', 'ml_pct_error']].describe())

       ml_abs_error  ml_pct_error
count   3096.000000   3096.000000
mean     184.399259     37.414772
std      413.651328    142.212387
min        0.056169      0.000660
25%       10.479469      0.393244
50%       37.507537      1.825544
75%      138.723372      9.161789
max     3798.922264   1956.134999


**Feature importance (interpretability only)**

Fit once on the full pooled set (not used for evaluation — this model saw all years, so it's for understanding what the model leans on, not for scoring).

In [40]:
full_train, _ = add_zscore_feature(master_vintages, master_vintages.head(1))
full_target = full_train[target_col] - full_train['Value']

full_model = lgb.LGBMRegressor(
    n_estimators=200,
    max_depth=4,
    learning_rate=0.05,
    random_state=42,
    verbosity=-1
)
full_model.fit(full_train[feature_cols_base], full_target)

importance = pd.Series(
    full_model.feature_importances_, index=feature_cols_base
).sort_values(ascending=False)
print(importance)

value_zscore         724
revision             389
horizon_months       355
Commodity            236
revision_pct         148
report_month          34
is_first_estimate      0
dtype: int32


# **EVALUATE AGAINST BASELINE**

Compare LOO MAE, pooled ML vs. raw WASDE estimate

In [41]:
ml_by_commodity_horizon = (
    oof.groupby(['Commodity', 'horizon_months'])
    .agg(mae_ml=('ml_abs_error', 'mean'), mape_ml=('ml_pct_error', 'mean'), n=('ml_abs_error', 'count'))
    .reset_index()
)

comparison = baseline_by_commodity_horizon_full.merge(
    ml_by_commodity_horizon, on=['Commodity', 'horizon_months'], suffixes=('', '_check')
)
comparison['mape_lift'] = comparison['mape_raw'] - comparison['mape_ml']

print(comparison.shape)
comparison.head(20)

(216, 9)


,Commodity,horizon_months,mae_raw,mape_raw,n,mae_ml,mape_ml,n_check,mape_lift
0,Corn,1.0,0.000000,0.000000,16,43.985212,0.332910,16,-0.332910
1,Corn,2.0,0.000000,0.000000,15,36.778167,0.271888,15,-0.271888
2,Corn,3.0,0.000000,0.000000,15,37.247005,0.275458,15,-0.275458
3,Corn,4.0,0.000000,0.000000,14,32.812114,0.244414,14,-0.244414
4,Corn,5.0,17.733333,0.127743,15,38.662284,0.284696,15,-0.156952
5,Corn,6.0,17.733333,0.127743,15,38.267307,0.282146,15,-0.154402
6,Corn,7.0,20.461538,0.147396,13,41.384328,0.302098,13,-0.154701
7,Corn,8.0,26.466667,0.188648,15,52.949754,0.385149,15,-0.196501
8,Corn,9.0,26.466667,0.188648,15,52.669659,0.384040,15,-0.195391
9,Corn,10.0,26.466667,0.188648,15,51.339077,0.373934,15,-0.185286


In [42]:
ml_mape_by_commodity = oof.groupby('Commodity')['ml_pct_error'].mean().sort_values()

comparison_overall = pd.DataFrame({
    'mape_raw': mape_by_commodity,
    'mape_ml': ml_mape_by_commodity
})
comparison_overall['lift'] = comparison_overall['mape_raw'] - comparison_overall['mape_ml']
comparison_overall = comparison_overall.sort_values('lift', ascending=False)

print(comparison_overall)
print('\nPooled overall MAPE — raw:', round(oof.merge(master_vintages[['Commodity','MarketYear','horizon_months','raw_pct_error']], on=['Commodity','MarketYear','horizon_months'])['raw_pct_error'].mean(), 4))
print('Pooled overall MAPE — ML: ', round(oof['ml_pct_error'].mean(), 4))

              mape_raw     mape_ml        lift
Commodity                                     
Soybean Oil   1.447467    1.455769   -0.008303
Sugar         1.618955    1.684974   -0.066019
Corn          1.134595    1.412629   -0.278034
Soybean Meal  1.435916    1.731137   -0.295221
Soybeans      1.331023    1.767746   -0.436723
Wheat         0.777000    1.805922   -1.028922
Rice (Rough)  1.429008   24.188682  -22.759674
Oats          2.602168   61.706727  -59.104560
Cotton        2.561545  240.979357 -238.417812

Pooled overall MAPE — raw: 1.5859
Pooled overall MAPE — ML:  37.4148


**Where the model fails: record-breaking years**

Cotton, Oats, and Rice show catastrophic per-row misses (up to ~2000% error) despite the pooled MAPE looking merely "bad" rather than broken. Inspecting the worst offenders:

In [43]:
commodity_range = master_vintages.groupby('Commodity')['final_value'].agg(['min', 'max'])

for c in ['Cotton', 'Oats', 'Rice (Rough)']:
    worst = oof[oof['Commodity'] == c].sort_values('ml_pct_error', ascending=False).head(5)
    print(f"--- {c} (historical final_value range: {commodity_range.loc[c, 'min']}-{commodity_range.loc[c, 'max']}) ---")
    print(worst[['MarketYear', 'horizon_months', 'final_value', 'ml_pred', 'ml_pct_error']].to_string(index=False))
    print()

--- Cotton (historical final_value range: 12.07-20.92) ---
MarketYear  horizon_months  final_value     ml_pred  ml_pct_error
   2022/23            21.0        14.47 -268.582734   1956.134999
   2016/17            22.0        17.17  313.861427   1727.964047
   2016/17            21.0        17.17  304.977942   1676.225638
   2023/24            18.0        12.07 -154.298461   1378.363390
   2023/24            20.0        12.07 -119.792735   1092.483308

--- Oats (historical final_value range: 40.0-93.0) ---
MarketYear  horizon_months  final_value    ml_pred  ml_pct_error
   2021/22            17.0         40.0 674.846975   1587.117437
   2021/22            18.0         40.0 659.590930   1548.977326
   2021/22            19.0         40.0 639.032890   1497.582226
   2021/22            20.0         40.0 595.428539   1388.571347
   2021/22            21.0         40.0 595.428539   1388.571347

--- Rice (Rough) (historical final_value range: 160.0-243.1) ---
MarketYear  horizon_months  final

Every one of the worst misses lands on a MarketYear whose `final_value` sits at (or right next to) that commodity's historical minimum. That's not a coincidence: `value_zscore` is recomputed per-fold from the *training* years only, so when the held-out year is a record low, its z-score falls outside anything the trees saw while splitting — leaf-average predictions can't extrapolate past their training range, so the reconstructed `Value + predicted_residual` overshoots wildly.

**Attempted mitigation:** clip the test-fold z-score to the training fold's own min/max before scoring, so record years get treated like the most extreme year the model actually trained on rather than an out-of-range value.

In [44]:
def add_zscore_feature_clipped(train_df, test_df):
    train_df, test_df = add_zscore_feature(train_df, test_df)
    zbounds = train_df.groupby('Commodity')['value_zscore'].agg(['min', 'max'])
    test_df = test_df.merge(zbounds, on='Commodity', how='left')
    test_df['value_zscore'] = test_df['value_zscore'].clip(lower=test_df['min'], upper=test_df['max'])
    test_df = test_df.drop(columns=['min', 'max'])
    return train_df, test_df

oof_clipped_rows = []

for train_idx, test_idx in logo.split(master_vintages, groups=groups):
    train_raw = master_vintages.iloc[train_idx].copy()
    test_raw = master_vintages.iloc[test_idx].copy()

    train_fold, test_fold = add_zscore_feature_clipped(train_raw, test_raw)
    train_target = train_fold[target_col] - train_fold['Value']

    model = lgb.LGBMRegressor(
        n_estimators=200, max_depth=4, learning_rate=0.05, random_state=42, verbosity=-1
    )
    model.fit(train_fold[feature_cols_base], train_target)

    test_fold = test_fold.copy()
    test_fold['ml_pred'] = test_fold['Value'] + model.predict(test_fold[feature_cols_base])
    oof_clipped_rows.append(test_fold[['Commodity', 'MarketYear', 'horizon_months', 'final_value', 'ml_pred']])

oof_clipped = pd.concat(oof_clipped_rows, ignore_index=True)
oof_clipped['ml_pct_error'] = ((oof_clipped['final_value'] - oof_clipped['ml_pred']) / oof_clipped['final_value']).abs() * 100

mape_ml_clipped_by_commodity = oof_clipped.groupby('Commodity')['ml_pct_error'].mean()

comparison_clipped = pd.DataFrame({
    'mape_raw': mape_by_commodity,
    'mape_ml': ml_mape_by_commodity,
    'mape_ml_clipped': mape_ml_clipped_by_commodity
})
print(comparison_clipped.sort_values('mape_raw'))
print('\nPooled MAPE — raw:              ', round(mape_by_commodity.mean(), 4), '(unweighted avg across commodities)')
print('Pooled MAPE — ML (unclipped):    ', round(oof['ml_pct_error'].mean(), 4))
print('Pooled MAPE — ML (clipped z):    ', round(oof_clipped['ml_pct_error'].mean(), 4))
print('\nMedian pct error — raw:          ', round(master_vintages['raw_pct_error'].median(), 4))
print('Median pct error — ML (unclipped):', round(oof['ml_pct_error'].median(), 4))

              mape_raw     mape_ml  mape_ml_clipped
Commodity                                          
Wheat         0.777000    1.805922         1.434700
Corn          1.134595    1.412629         1.413083
Soybeans      1.331023    1.767746         1.767746
Rice (Rough)  1.429008   24.188682        14.373243
Soybean Meal  1.435916    1.731137         1.730012
Soybean Oil   1.447467    1.455769         1.481026
Sugar         1.618955    1.684974         1.685769
Cotton        2.561545  240.979357       248.785782
Oats          2.602168   61.706727        33.082247

Pooled MAPE — raw:               1.5931 (unweighted avg across commodities)
Pooled MAPE — ML (unclipped):     37.4148
Pooled MAPE — ML (clipped z):     33.9726

Median pct error — raw:           0.2
Median pct error — ML (unclipped): 1.8255


# **RESULTS WRITE UP**

**Verdict: the pooled GBM does not beat the raw WASDE baseline. Don't ship it.**

- **Six of nine commodities land close to the baseline but never below it** — Soybean Oil, Sugar, Corn, Soybean Meal, Soybeans, and Wheat sit within ~0.01–1.8 MAPE points of the raw estimate. The model isn't finding exploitable signal beyond what USDA's own live estimate already encodes; at best it's roughly matching it, and at every horizon checked so far it's slightly worse.
- **Three commodities fail badly**: Cotton (137–249% MAPE), Oats (33–62%), Rice (14–24%). Every catastrophic miss traces to a held-out MarketYear sitting at that commodity's historical min — a record year pushes `value_zscore` outside the range the trees trained on, and leaf-average predictions can't extrapolate, so `Value + predicted_residual` overshoots by 100s of percent.
- **Clipping the z-score to the training fold's range partially helps** (Oats 62%→33%, Rice 24%→14%) but doesn't fix Cotton, and doesn't change the top-line conclusion.
- **The median row (1.8% ML vs 0.2% raw) is worse than baseline too**, not just the outlier-driven mean — so this isn't purely a tail problem. For short horizons especially (1–4 months out), WASDE's own number is already essentially exact (raw MAE = 0), and the model just adds noise where there's nothing to correct.

**Why this happened, most likely:**
1. **Too little data per commodity.** 16 market years per commodity, pooled into ~3,100 rows, isn't enough for a tree ensemble to learn commodity-specific revision dynamics — especially for small-magnitude commodities like Cotton, where a routine absolute revision is a huge *percentage* swing.
2. **Tree ensembles can't extrapolate**, and this dataset is short and trending by construction (16 non-independent market years, several of them record highs/lows for low-sample commodities). Any feature or target built from per-fold statistics (z-score) inherits that fragility.
3. **Pooling assumed shared structure that may not exist.** `Commodity` shows up as the 3rd-4th most important feature both times — the model leans on it heavily, which suggests commodity-specific behavior dominates over any cross-commodity revision pattern the pooling was meant to exploit.

**Recommended next steps, in order of effort:**
- **Keep the raw WASDE baseline (`raw_wasde_baseline.csv`) as the production estimate.** Nothing here beats it.
- **Compare against the existing bridge-regression (OLS) approach per-commodity** (see `time_ser_approach.ipynb`) rather than a single pooled tree model — a linear model on `Value` can extrapolate through its coefficient in a way trees structurally cannot, which may explain why that approach was tried first for Corn.
- **If continuing with trees**, add a guardrail: fall back to the raw `Value` (zero predicted residual) whenever the current z-score falls outside the training fold's observed range, rather than trusting an extrapolated leaf prediction.
- **Don't pool small-magnitude, low-sample commodities (Cotton, Oats, Rice) with the rest** — either model them separately with tighter regularization, or exclude them from the pooled model and keep using the raw baseline for them specifically.
- **Hyperparameters here were never tuned** (`n_estimators=200, max_depth=4, learning_rate=0.05` chosen arbitrarily) — not worth tuning until the fundamental data-volume and extrapolation issues above are addressed, since tuning won't fix either.